File: exp_contour.m
% How to plot experimental data directivity contours

expdir = '/projectnb/turbomac/arozman/projects/ULI/experimental_data/16degPitch_90degYaw_10mps_AutoSpec.mat';
micsdir = '/projectnb/turbomac/arozman/projects/ULI/experimental_data/Xc_U10.txt';

% Load data from MAT file directly into workspace
% Variables loaded: freq, broadband_spectra, rpm, tonal_spectra, total_spectra
load(expdir);

fq = squeeze(freq);
rpms = squeeze(rpm);
bbn = broadband_spectra;   % Size: [4097, 251, 21] (freq x mic x rpm)
tonal = tonal_spectra;     % Size: [4097, 251, 21]
total = total_spectra;     % Size: [4097, 251, 21]

xyz = readmatrix(micsdir);

%% Filter by BPF, remove NaNs
rpm_ix = length(rpms);   % 4000 RPM (last RPM entry)
BPF = rpms(rpm_ix) / 60 * 5;

% Index in fq array of BPF frequency
[~, ix] = min(abs(fq - BPF));

% Microphones with non-NaN values at BPF for the target RPM
mask = ~isnan(squeeze(tonal(ix, :, rpm_ix)));
if isrow(mask)
    mask = mask';
end

% Explicitly bad microphone indices (0-indexed [17, 155, 219] -> 1-indexed [18, 156, 220])
bad_mics = [18, 156, 220];
mask(bad_mics) = false;

% Remove invalid mic entries from xyz and spectral arrays
xyz = xyz(mask, :);
tonal = tonal(:, mask, :);
total = total(:, mask, :);
bbn = bbn(:, mask, :);

%% Interp PSD on grid & acoustic processing
[X, Y] = meshgrid(linspace(-2, 2, 100), linspace(-1, 1, 50));

tonal_bpf = squeeze(tonal(ix, :, rpm_ix));
Z = griddata(xyz(:,1), xyz(:,2), tonal_bpf, X, Y, 'linear');

df = fq(2) - fq(1);

% SPL (dB) is multiplied by bandwidth
Z_spectrum = 10 * log10( Z * df / (2e-5)^2 );

% PSD (dB/Hz) is not.
Z_PSD = 10 * log10( Z / (2e-5)^2 );

% Values for contour colorbar
levels = linspace(65, 75, 75);

%% Plotting

% 1. Plot SPL
figure;
ax1 = gca;
contourf(X, Y, clamp_levels(Z_spectrum, levels), levels, 'LineStyle', 'none');
colormap(jet);
clim([levels(1), levels(end)]);
format_fig(ax1, levels, 'SPL (dB)');
title(ax1, sprintf('Experiment 4000RPM Hover SPL, df=%g', fq(2)), 'FontSize', 14);

% 2. Plot PSD
figure;
ax2 = gca;
contourf(X, Y, clamp_levels(Z_PSD, levels), levels, 'LineStyle', 'none');
colormap(jet);
clim([levels(1), levels(end)]);
format_fig(ax2, levels, 'PSD (dB/sqrt(Hz))');
title(ax2, sprintf('Experiment 4000RPM Hover PSD, df=%g', fq(2)), 'FontSize', 14);

% 3. Plot PSD averaged within +/- 70Hz of BPF
freq_mask_70 = (fq >= (BPF - 70)) & (fq <= (BPF + 70));
tonal_last_rpm = squeeze(tonal(:, :, rpm_ix)); % Size: 4097 x N_mics
tonal_avg_70Hz = mean(tonal_last_rpm(freq_mask_70, :), 1)';
Z_avg_70Hz = griddata(xyz(:, 1), xyz(:, 2), tonal_avg_70Hz, X, Y, 'linear');
Z_PSD_70Hz = 10 * log10(Z_avg_70Hz / (2e-5)^2);

figure;
ax3 = gca;
contourf(X, Y, clamp_levels(Z_PSD_70Hz, levels), levels, 'LineStyle', 'none');
colormap(jet);
clim([levels(1), levels(end)]);
format_fig(ax3, levels, 'PSD (dB/Hz)');
title(ax3, sprintf('Experiment 4000RPM Hover PSD (Avg +/- 70Hz), df=%g', fq(2)), 'FontSize', 14);

% 4. Plot PSD averaged within +/- 10Hz of BPF
freq_mask_10 = (fq >= (BPF - 10)) & (fq <= (BPF + 10));
tonal_avg_10Hz = mean(tonal_last_rpm(freq_mask_10, :), 1)';
Z_avg_10Hz = griddata(xyz(:, 1), xyz(:, 2), tonal_avg_10Hz, X, Y, 'linear');
Z_PSD_10Hz = 10 * log10(Z_avg_10Hz / (2e-5)^2);

figure;
ax4 = gca;
contourf(X, Y, clamp_levels(Z_PSD_10Hz, levels), levels, 'LineStyle', 'none');
colormap(jet);
clim([levels(1), levels(end)]);
format_fig(ax4, levels, 'PSD (dB/Hz)');
title(ax4, sprintf('Experiment 4000RPM Hover PSD (Avg +/- 10Hz), df=%g', fq(2)), 'FontSize', 14);

% 5. Plot broadband noise (BBN) integrated from 500 to 20000 Hz
bbn_mask = (fq >= 500) & (fq <= 20000);
bbn_last_rpm = squeeze(bbn(:, :, rpm_ix)); % Size: 4097 x N_mics
bbn_int = sum(bbn_last_rpm(bbn_mask, :) * df, 1)';
Z_bbn = griddata(xyz(:, 1), xyz(:, 2), bbn_int, X, Y, 'linear');
Z_bbn_SPL = 10 * log10(Z_bbn / (2e-5)^2);

levels_bbn = linspace(90, 100, 75);
figure;
ax5 = gca;
contourf(X, Y, clamp_levels(Z_bbn_SPL, levels_bbn), levels_bbn, 'LineStyle', 'none');
colormap(jet);
clim([levels_bbn(1), levels_bbn(end)]);
format_fig(ax5, levels_bbn, 'BBN SPL (dB)');
title(ax5, 'Experiment 4000RPM Broadband Noise SPL (500-20000 Hz)', 'FontSize', 14);

%% Helper Functions

% Clamp data values to level bounds so contourf fills out-of-bounds regions (extend='both')
function Zc = clamp_levels(Z, levels)
    Zc = Z;
    Zc(Z < levels(1)) = levels(1);
    Zc(Z > levels(end)) = levels(end);
end

% Figure formatting helper
function format_fig(ax, levels, cb_title)
    xlabel(ax, 'X (m)', 'FontSize', 14);
    ylabel(ax, 'Y (m)', 'FontSize', 14);
    set(ax, 'FontSize', 10);
    axis(ax, 'equal');
    
    cbar = colorbar(ax);
    ylabel(cbar, cb_title, 'FontSize', 14);
    ticks = linspace(levels(1), levels(end), 10);
    cbar.Ticks = ticks;
    cbar.TickLabels = arrayfun(@(x) sprintf('%.2f', x), ticks, 'UniformOutput', false);

    ylim(ax, [-1, 1]);
    xlim(ax, [-2, 2]);
end

File: exp_contour.py
# -*- coding: utf-8 -*-
"""
How to plot experimental data directivity contours

"""

# import numpy as np
# import scipy.io as sio
import numpy as np
import matplotlib.pyplot as plt
from mpl_toolkits.axes_grid1 import make_axes_locatable
import h5py
from scipy.interpolate import griddata

expdir="/projectnb/turbomac/arozman/projects/ULI/experimental_data/16degPitch_90degYaw_10mps_AutoSpec.mat"
micsdir="/projectnb/turbomac/arozman/projects/ULI/experimental_data/Xc_U10.txt"

# load data
file =  h5py.File(expdir)

print(file.keys())

fq = np.squeeze(file['freq'][:])
bbn = file['broadband_spectra'][:]
rpms = np.squeeze(file['rpm'][:])
tonal = file['tonal_spectra'][:]
total = file['total_spectra'][:]

xyz = np.loadtxt(micsdir,delimiter=",")

def format_fig(fig, ax, CS, levels, title):
    ax.set_xlabel('X (m)', fontsize=14)
    ax.set_ylabel('Y (m)', fontsize=14)
    ax.tick_params(axis='both', labelsize=10)
    plt.axis('equal')

    divider = make_axes_locatable(ax)
    cax = divider.append_axes("right", size="5%", pad=0.1)
    
    ticks = np.linspace(levels[0], levels[-1], 10).tolist()
    
    cbar = fig.colorbar(CS, cax=cax, ticks=ticks)
    cbar.ax.set_ylabel(title, fontsize=14)
    cbar.ax.set_yticklabels([f'{x:.2f}' for x in ticks])

    ax.set_ylim((-1.,1.))
    ax.set_xlim((-2.,2.))
# %% Filter by BPF, remove NaNs
rpm_ix=-1   # 4000 RPM
BPF = rpms[rpm_ix]/60*5

# index in fq array of BPF frequency
ix = np.argmin(abs(fq-BPF))

# where there are not NaNs at the BPF 
mask = ~np.isnan(tonal[-1,:,ix])

# also remove the explicitly bad mics
bad_mics = [17, 155, 219]

mask[bad_mics]=False

# remove those values from xyz and noise arrays
xyz = xyz[mask,:]
tonal = tonal[:,mask,:]
total = total[:,mask,:]
bbn = bbn[:,mask,:]


# %% Interp PDD on grid & acoustic processing
X, Y, = np.meshgrid(np.linspace(-2,2,100), np.linspace(-1,1,50))
Z = griddata((xyz[:,0], xyz[:,1]), tonal[-1,:,ix], (X,Y), method='linear')

df = fq[1] - fq[0]

# SPL (dB) is multiplied by bandwidth
Z_spectrum = 10*np.log10( Z * df / (2e-5)**2 )

# PSD (dB/Hz) is not.
Z_PSD = 10*np.log10( Z / (2e-5)**2 )

# values for contour colorbar
levels = np.linspace(65,75,75)

# %% plotting
# plot SPL
fig, ax = plt.subplots(layout='constrained')
CS = ax.contourf(X, Y, Z_spectrum, levels, extend='both', cmap="jet")
format_fig(fig, ax, CS, levels, 'SPL (dB)')
ax.set_title(f"Experiment 4000RPM Hover SPL, df={fq[1]}", fontsize=14)

# %%
# plot PSD
fig, ax = plt.subplots(layout='constrained')
CS = ax.contourf(X, Y, Z_PSD, levels, extend='both', cmap="jet")
format_fig(fig, ax, CS, levels, 'PSD (dB/sqrt(Hz))')
ax.set_title(f"Experiment 4000RPM Hover PSD, df={fq[1]}", fontsize=14)

# %%
# plot PSD averaged within +/- 70Hz of BPF
freq_mask = (fq >= (BPF - 70)) & (fq <= (BPF + 70))
tonal_avg_70Hz = np.mean(tonal[-1][:, freq_mask], axis=1)
Z_avg_70Hz = griddata((xyz[:, 0], xyz[:, 1]), tonal_avg_70Hz, (X, Y), method='linear')
Z_PSD_70Hz = 10 * np.log10(Z_avg_70Hz / (2e-5)**2)

fig, ax = plt.subplots(layout='constrained')
CS = ax.contourf(X, Y, Z_PSD_70Hz, levels, extend='both', cmap="jet")
format_fig(fig, ax, CS, levels, 'PSD (dB/Hz)')
ax.set_title(f"Experiment 4000RPM Hover PSD (Avg +/- 70Hz), df={fq[1]}", fontsize=14)

# %%
# plot PSD averaged within +/- 10Hz of BPF
freq_mask = (fq >= (BPF - 10)) & (fq <= (BPF + 10))
tonal_avg_70Hz = np.mean(tonal[-1][:, freq_mask], axis=1)
Z_avg_70Hz = griddata((xyz[:, 0], xyz[:, 1]), tonal_avg_70Hz, (X, Y), method='linear')
Z_PSD_70Hz = 10 * np.log10(Z_avg_70Hz / (2e-5)**2)

fig, ax = plt.subplots(layout='constrained')
CS = ax.contourf(X, Y, Z_PSD_70Hz, levels, extend='both', cmap="jet")
format_fig(fig, ax, CS, levels, 'PSD (dB/Hz)')
ax.set_title(f"Experiment 4000RPM Hover PSD (Avg +/- 10Hz), df={fq[1]}", fontsize=14)

# %%
# plot broadband noise (BBN) integrated from 500 to 20000 Hz
bbn_mask = (fq >= 500) & (fq <= 20000)
bbn_int = np.sum(bbn[-1][:, bbn_mask] * df, axis=1)
Z_bbn = griddata((xyz[:, 0], xyz[:, 1]), bbn_int, (X, Y), method='linear')
Z_bbn_SPL = 10 * np.log10(Z_bbn / (2e-5)**2)

levels_bbn = np.linspace(90, 100, 75)
fig, ax = plt.subplots(layout='constrained')
CS = ax.contourf(X, Y, Z_bbn_SPL, levels_bbn, extend='both', cmap="jet")
format_fig(fig, ax, CS, levels_bbn, 'BBN SPL (dB)')
ax.set_title("Experiment 4000RPM Broadband Noise SPL (500-20000 Hz)", fontsize=14)

plt.show()

